polygeist/lib/Dialect/Polygeist/Transforms/FunctionSpecialization.cpp

```diff
@@ -0,0 +1,281 @@
+using AccessorPtrType = VersionConditionBuilder::AccessorPtrType;
+using AccessorPtrPairType = VersionConditionBuilder::AccessorPtrPairType;
+
+static llvm::cl::opt<unsigned> functionSpecializationAccessorLimit(
+    "function-specialization-accessor-limit", llvm::cl::init(5),
+    llvm::cl::desc(
+        "Maximum number of accessors allowed for function specialization"));
+
+//===----------------------------------------------------------------------===//
+// Helper Functions
+//===----------------------------------------------------------------------===//
+
+/// Returns true if \p type is 'memref<?x!sycl.accessor>', and false otherwise.
+static bool isValidMemRefType(Type type) {
+  auto mt = dyn_cast<MemRefType>(type);
+  bool isMemRefWithExpectedShape =
+      (mt && mt.hasRank() && (mt.getRank() == 1) &&
+       ShapedType::isDynamic(mt.getShape()[0]) && mt.getLayout().isIdentity());
+  if (!isMemRefWithExpectedShape)
+    return false;
+
+  return isa<sycl::AccessorType>(mt.getElementType());
+}
+
+/// Returns true if \p arg is a candidate argument. Currently, all arguments
+/// with valid memref type determined by isValidMemRefType is a candidate
+/// argument, i.e., 'memref<?x!sycl.accessor>`.
+static bool isCandidateArg(Value arg) {
+  if (isValidMemRefType(arg.getType())) {
+    assert(isa<AccessorPtrType>(arg) &&
+           "Expecting valid memref type to be AccessorPtrType");
+    return true;
+  }
+  return false;
+}
+
+/// Populate \p candArgs with candidate arguments from \p call.
+static void
+collectCandidateArguments(CallOpInterface call,
+                          SmallVectorImpl<AccessorPtrType> &candArgs) {
+  assert(candArgs.empty() && "Expecting empty candArgs");
+  for (Value arg : call.getArgOperands())
+    if (isCandidateArg(arg))
+      candArgs.push_back(cast<AccessorPtrType>(arg));
+}
+
+/// Returns the cloned version of \p func.
+static FunctionOpInterface cloneFunction(FunctionOpInterface func) {
+  FunctionOpInterface clonedFunc = func.clone();
+  std::string newFnName = (func.getName() + ".specialized").str();
+#ifndef NDEBUG
+  ModuleOp module = func->getParentOfType<ModuleOp>();
+  module->walk([newFnName](FunctionOpInterface func) {
+    assert(func.getName() != newFnName &&
+           "Expecting new function name to be unique");
+  });
+#endif // NDEBUG
+  clonedFunc.setName(newFnName);
+  OpBuilder builder(func);
+  builder.insert(clonedFunc);
+  privatize(clonedFunc);
+  return clonedFunc;
+}
+
+/// Add attribute 'sycl.inner.disjoint' to arguments of \p func that satisfy \p
+/// predicate.
+static void setInnerDisjointAttribute(FunctionOpInterface func,
+                                      std::function<bool(Value)> predicate) {
+  constexpr StringRef innerDisjointAttrName = "sycl.inner.disjoint";
+  for (unsigned i = 0; i < func.getNumArguments(); ++i)
+    if (predicate(func.getArgument(i)))
+      func.setArgAttr(i, innerDisjointAttrName,
+                      UnitAttr::get(func->getContext()));
+}
```

Comment:
```suggestion
/// Add attribute 'sycl.inner.disjoint' to arguments of \p func that satisfy \p
/// predicate.
template <typename Pred, typename = std::enable_if_t<std::is_invocable_r_v<bool, Pred, Value>>>
static void setInnerDisjointAttribute(FunctionOpInterface func, Pred predicate) {
  constexpr StringRef innerDisjointAttrName = "sycl.inner.disjoint";
  for (unsigned i = 0; i < func.getNumArguments(); ++i)
    if (predicate(func.getArgument(i)))
      func.setArgAttr(i, innerDisjointAttrName,
                      UnitAttr::get(func->getContext()));
}
```
Keep check of function signature while making more generic

Reply:
curious, if we keep checking function signature, why is it more generic?

Reply:
Because we can pass a function pointer, an `std::function<...>`, `llvm::function_ref<...>`, a lambda, a function object, etc. with no conversion.

Reply:
oic, good to know, thanks